src/GNDStk/gnds-tree-node.hpp

```diff
@@ -0,0 +1,444 @@
+
+// -----------------------------------------------------------------------------
+// Node (templated)
+// Then node = Node<>
+// -----------------------------------------------------------------------------
+
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+class Node {
+   // for internal use
+   static bool static_found;
+
+   using pair = std::pair<std::string,std::string>;
+   using sptr = std::shared_ptr<Node>;
+
+public:
+
+   // Simple node for our tree structure:
+   //    name
+   //    metadata
+   //    children
+   std::string name;
+   MCON<pair,std::allocator<pair>> metadata;
+   CCON<sptr,std::allocator<sptr>> children;
+
+   // normalize
+   void normalize();
+
+   // push metadatum
+   pair &push(const std::string &key, const std::string &value)
+   {
+      metadata.push_back(std::make_pair(key,value));
+      return metadata.back();
+   }
+
+   pair &push(const pair &m)
+   {
+      metadata.push_back(m);
+      return metadata.back();
+   }
+
+   template<class T>
+   pair &push(const std::string &key, const T &value)
+   {
+      std::ostringstream oss;
+      oss << value;
+      return push(key,oss.str());
+   }
+
+   template<class T>
+   pair &push(const std::pair<std::string,T> &m)
+   {
+      std::ostringstream oss;
+      oss << m.second;
+      return push(pair(m.first,oss.str()));
+   }
+
+   // push child
+   Node &push(const std::string &name = "")
+   {
+      children.push_back(std::make_shared<Node<MCON,CCON>>());
+      children.back()->name = name;
+      return *children.back();
+   }
+
+   // write
+   bool write(const char * const file, const int level = 0) const;
+   bool write(const std::string &file, const int level = 0) const
+   {
+      return write(file.c_str(),level);
+   }
+   std::ostream &write(std::ostream &, const int level = 0) const;
+
+   // leaf?
+   bool leaf() const
+   {
+      return children.size() == 0;
+   }
+
+
+   // ------------------------
+   // meta()
+   // ------------------------
+
+   // for string
+   const std::string &meta(
```

Comment:
I lean towards shorter names, in the traditional C/C++ style. Like int instead of integer, char instead of character, and so on and so forth in the language itself. Here, additionally, there's a bit of a grammatical issue. "Data" (and "metadata") is plural, even though most humans, in both speaking voice and writing voice, use "data" as singular. To be pedantic about it, meta() doesn't return *metadata*, it return *a* *metadatum*. Just like child(), just below it, return *a child*, not several children. We can have some offline discussion about names. Or maybe do as advertising agencies do - put a few options out to consumers (users), and see how positively or negatively they respond to this or that. Ultimately, my approach has always been to give users what *they* perceive to be a simple, uncluttered, and logical API.